pkg/devicetwin/service.go

```diff
@@ -0,0 +1,429 @@
+				SnapListHash:       versionedPublishSnaps.Result.SnapListHash,
+				InstalledSnapsHash: versionedPublishSnaps.Result.InstalledSnapsHash,
+				LastRefresh:        time.Now(),
+			})
+
+			if tx.Error != nil {
+				return fmt.Errorf("error trying to update health hashes for %s: %w", healthHashes.DeviceID, tx.Error)
+			}
+
+			// build a message payload for ActionResponse
+			err = srv.sendLegacyActionHandlerUnversionedPayload(clientID, msg, versionedMessage)
+			if err != nil {
+				return fmt.Errorf("error trying to send unversioned payload to legacy action handler: %w", err)
+			}
+			return nil
+		default:
+			logger.Errorf("Received a versioned action message we cannot handled: %s", versionedMessage.Action)
+			return fmt.Errorf("error trying to handle versioned action message type %s", versionedMessage.Action)
+		}
+	default:
+		err := fmt.Errorf("received a versioned action message with a version we don't expect: %s", versionedMessage.Version)
+		logger.Error(err)
+		return err
+
+	}
+
+	return nil
+}
+
+func (srv *Service) healthMessageHandler(msg MQTT.Message) error {
+	// After sending the message to the legacy handler, do our own processing
+	clientID := getClientID(msg)
+	logger.Printf("Health update from %s", clientID)
+
+	// Parse the body
+	h := messages.Health{}
+	if err := json.Unmarshal(msg.Payload(), &h); err != nil {
+		return fmt.Errorf("error trying to unmarshal health message payload: %w", err)
+	}
+
+	// Check that the client ID matches
+	if clientID != h.DeviceId {
+		return fmt.Errorf("client/device ID mismatch: %s from the topic and %s from the message; discarding", clientID, h.DeviceId)
+	}
+
+	logger.Infof("Received health message %+v, sending to iot-devicetwin", msg)
+	srv.legacyHealthChan <- msg
+
+	var healthMessage messages.Health
+	err := json.Unmarshal(msg.Payload(), &healthMessage)
+	if err != nil {
+		return fmt.Errorf("failed unmarshaling a health message: %w", err)
+	}
+
+	if healthMessage.InstalledSnapsHash == "" && healthMessage.SnapListHash == "" {
+		// No error, could be just a device that does not support this yet
+		logger.Infof("Received health message from %s but did not have hashes", healthMessage.DeviceId)
+		return nil
+	}
+
+	var healthHashes models.HealthHash
+	tx := srv.db.Find(&healthHashes, &models.HealthHash{DeviceID: healthMessage.DeviceId})
+	if tx.Error != nil {
+		return fmt.Errorf("error trying to find health hash for %s: %w", healthMessage.DeviceId, tx.Error)
+	}
+
+	if tx.RowsAffected == 0 {
+		srv.db.Create(&models.HealthHash{
+			LastRefresh:        time.Now(),
+			OrgID:              healthMessage.OrgId,
+			DeviceID:           healthMessage.DeviceId,
+			SnapListHash:       healthMessage.SnapListHash,
+			InstalledSnapsHash: healthMessage.InstalledSnapsHash,
+		})
+
+		return nil
+	}
+
+	// Do the health hashes match?
+	if healthHashes.SnapListHash == healthMessage.SnapListHash && healthHashes.InstalledSnapsHash == healthMessage.InstalledSnapsHash {
```

Comment:
Is the plan to for `SnapListHash` included snap config changes? If so, could that serve to keep devicetwin synchronized with devices?

Reply:
No, that's InstalledSnapsHash. The names could probably use improvement, open to suggestions.